framalytics/fram.py

```diff
@@ -216,6 +217,71 @@ def generate_connection_traversal_count(self, output_function):
 
 
 
+    def highlight_data(self, data):
+        """
+        Highlights the connections of all bezier curves which data instances traverse.
+        The color of the connection indicates the intensity of its usage.
+
+        :param data: A dataframe where the columns are function names, and rows are instances.
+        The values of each column should be 0 (absent) or 1 (present)
+
+        :return: None. A highlighted FRAM model.
+        """
+        total_instances = len(data)  # Total number of rows / instances from the dataframe.
+
+        for i in self.get_connections():
+            # Stores current connection name
+            connection_default = i
+
+            # Splits connection to get information
+            connection = i.split("|")
+
+            # Get output and input function ID's
+            outputFn = connection[0]
+            toFn = connection[2]
+
+            # Stores the names of the functions.
+            outputFn_name = self.functions_by_id.get(outputFn)
+            toFn_name = self.functions_by_id.get(toFn)
+
+            # Begins iterating through the dataframe to find the number of times a path is traversed
+            # This is based on if both functions in an instance have a value of 1.
+            for row in range(len(data[outputFn_name])):
+                outputFn_data_value = data[outputFn_name][row]
+                toFn_data_value = data[toFn_name][row]
+
+                if ((outputFn_data_value == 1) and (toFn_data_value == 1)):
+                    connection_value = self.connections.get(connection_default)
+                    connection_value += 1
+                    self.connections.update({connection_default: connection_value})
+
```

Comment:
We can avoid this inner loop by using pandas slicing and len().

`connection_value = len(data[(data[outputFn_name] == 1) & (data[toFn_name] == 1)])`